### PR TITLE
rosie disco: add libsecret password manager

### LINKED ISSUE #2168: rosie disco client: support GObject equivalent of GnomeKeyring
Example can be found here: https://developer.gnome.org/libsecret/unstable/py-examples.html

Looks like it should be relatively straightforward to program up something for GObject libsecret. :secret:

### PR DESCRIPTION
**Pull request by sadielbartholomew**
Closes #2168.

_Note:_ ~~currently for preliminary feedback only, since I am as yet unable to test for functionality due to lack of module availability for ``cherrypy`` & ``sqlaclhemy`` on the present environment installed on the internal RHEL 7 boxes.~~ ~~Awaiting an SRS account to test with.~~ Open for review.

### CODE REVIEW

File: lib/python/rosie/ws_client_auth.py [hunk 1]
```diff
@@ -236,16 +248,74 @@ def store_password(self, scheme, host, username, password):
...
+
+    """Password management with libsecret."""
+
+    @classmethod
+    def usable(cls):
+        """Can this store be used?"""
+        if gi_flag:
+            return True
+        else:
+            return False
```
[the remark this answers is not in the document]
Comment (sadielbartholomew) [reply lib/python/rosie/ws_client_auth.py]: Very good point! Confused as to why I didn't do that instead...

File: lib/python/rosie/ws_client_auth.py [hunk 2]
```diff
@@ -236,16 +248,74 @@ def store_password(self, scheme, host, username, password):
...
+        self.schema = Secret.Schema.new(
+            "org.rosie.disco.Store", Secret.SchemaFlags.NONE, self.attributes)
+
+    def clear_password(self, scheme, host, username):
+        """Remove the password from the cache."""
+        try:
+            if (scheme, host, username) in self.attributes:
+                key, value = self.attributes[(scheme, host, username)]
+                Secret.password_clear(self.schema, (key, value), None,
+                                      Secret.password_clear_finish())
```
[the remark this answers is not in the document]
Comment (sadielbartholomew) [reply lib/python/rosie/ws_client_auth.py]: Thanks, will have a look shortly.

File: lib/python/rosie/ws_client_auth.py [hunk 3]
```diff
@@ -236,16 +248,74 @@ def store_password(self, scheme, host, username, password):
...
+            return True
+        else:
+            return False
+
+    def __init__(self):
+        self.attributes = {
+            "user": Secret.SchemaAttributeType.STRING,
+            "protocol": Secret.SchemaAttributeType.STRING,
+            "server": Secret.SchemaAttributeType.STRING,
+        }
```
[the remark this answers is not in the document]
Comment (sadielbartholomew) [reply lib/python/rosie/ws_client_auth.py]: I see, thanks for picking up on this. I wrote this fairly quickly with the plan to immediately go on to straighten out these kind of issues via testing within an environment with the necessary dependencies for 'libsercret' - shame I cannot yet do this :(.

**Comment by sadielbartholomew** [comment]
(Travis CI original failure unrelated to changes - see #2172 - & now resolved via rebase to merged master. PR now passes.)

**Comment by sadielbartholomew** [comment]
I returned to this after the promise that ``cherrypy`` & ``sqlaclhemy`` were now available on (some of) the RHEL7 boxes for testing, but can't seem to access them still :(. I'll seek advice on this next week & hopefully get this PR tested & finalised. For the moment, initial feedback has been addressed.

**Comment by matthewrmshin** [comment]
Test using Ubuntu 16.04 LTS with GNOME Shell 3.18.5 requires the following change.

```diff
diff --git a/lib/python/rosie/ws_client_auth.py b/lib/python/rosie/ws_client_auth.py
index ef05d478..49481589 100644
--- a/lib/python/rosie/ws_client_auth.py
+++ b/lib/python/rosie/ws_client_auth.py
@@ -38,7 +38,7 @@ import socket
 try:
     from gi import require_version, pygtkcompat
     require_version('Gtk', '3.0')  # For GTK+ >=v3 use PyGObject; v2 use PyGTK
-    require_version('Secret', '1.0')
+    require_version('Secret', '1')
     from gi.repository import Secret
     gi_flag = True
 except ImportError:
@@ -259,46 +259,37 @@ class LibsecretStore(object):
         return bool(gi_flag)
 
     def __init__(self):
-        self.template = {
-            "protocol": scheme,
-            "server": host,
-            "user": username
-        }
         # Attributes must be explicitly defined; not managed by the schema
-        self.attributes = dict((key, Secret.SchemaAttributeType.STRING) for
-                               key in self.template)
+        self.attributes = {
+            "protocol": Secret.SchemaAttributeType.STRING,
+            "server": Secret.SchemaAttributeType.STRING,
+            "user": Secret.SchemaAttributeType.STRING}
         self.schema = Secret.Schema.new(
             "org.rosie.disco.Store", Secret.SchemaFlags.NONE, self.attributes)
 
     def clear_password(self, scheme, host, username):
         """Remove the password from the cache."""
-        try:
-            if all(attr in self.attributes.values for attr in
-                    self.template.values):
-                Secret.password_clear_sync(self.schema, self.template, None)
-        except Secret.SECRET_ERROR_NO_SUCH_OBJECT:
-            pass
+        template = {"protocol": scheme, "server": host, "user": username}
+        Secret.password_clear_sync(self.schema, template, None)
 
     def find_password(self, scheme, host, username):
         """Return the password of username@root."""
+        template = {"protocol": scheme, "server": host, "user": username}
         try:
-            password = Secret.password_lookup_sync(self.schema, self.template,
-                                                   None)
+            password = Secret.password_lookup_sync(self.schema, template, None)
         except Secret.SECRET_ERROR_PROTOCOL:
             return
-        if password is None:
-            raise Secret.SECRET_ERROR_NO_SUCH_OBJECT
         return password
 
     def store_password(self, scheme, host, username, password):
         """Return the password of username@root."""
         self.clear_password(scheme, host, username)
+        template = {"protocol": scheme, "server": host, "user": username}
         try:
             Secret.password_store_sync(
-                self.schema, self.template, Secret.COLLECTION_DEFAULT, host,
+                self.schema, template, Secret.COLLECTION_DEFAULT, host,
                 password, None)
-        except (Secret.SECRET_ERROR_NO_SUCH_OBJECT,
-                Secret.SECRET_ERROR_PROTOCOL):
+        except Secret.SECRET_ERROR_PROTOCOL:
             pass
 
 
```

**Comment by sadielbartholomew** [comment]
Thank you @matthewrmshin. I was a bit alarmed to see so many line changes were necessary but it seems predominantly because ``scheme``, ``host`` & ``username`` were used before being defined for the ``template`` dictionary. Oops! That's why testing is helpful :)

The required modules are now available on the RHEL7 boxes so I will quickly do my own tests before opening for review.

**Comment by sadielbartholomew** [comment]
Now in fact waiting on an SRS account which I also need to test this, but as soon as I have one I will be able to test quickly and open this for review.

**Comment by sadielbartholomew** [comment]
I obtained an SRS account & tested this on a RHEL7 box (eld516). It seemed to work fine. @matthewrmshin & @oliver-sanders: now open for review. Thanks.

File: lib/python/rosie/ws_client_auth.py [hunk 4]
```diff
@@ -22,22 +22,35 @@
...
+
 import socket
-import sys
-from urlparse import urlparse
+try:
+    from gi import require_version, pygtkcompat
+    require_version('Gtk', '3.0')  # For GTK+ >=v3 use PyGObject; v2 use PyGTK
+    require_version('Secret', '1')
+    from gi.repository import Secret
+    gi_flag = True
```
Comment (oliver-sanders) [inline lib/python/rosie/ws_client_auth.py]: This is a constant so should really use uppercase.

Comment (sadielbartholomew) [reply lib/python/rosie/ws_client_auth.py]: Good point! Will get it changed now.

**Pull request merged**